### ISSUE TITLE
 Isabelle task 4 

### ISSUE THREAD

**Issue by oflefe** [johnwickerson/HSV#7]
Hi,  
I do not really understand how to go about task four. How would I descend in the circuit hierarchy to check every location for consecutive not gates. Because, checking if a circuit is NOT NOT does not mean checking if it contains a section with NOT NOT if I am getting it correctly.

**Comment by MattWindsor91** [comment]
Hi Efe,

> I do not really understand how to go about task four. How would I descend in the circuit hierarchy to check every location for consecutive not gates.

As the task is to make a function that you’ll eventually use to check whether `opt_NOT` (which _removes_ consecutive `NOT` gates) is correct, it might help to look at how `opt_NOT` is defined and see if that helps you.

> Because, checking if a circuit is NOT NOT does not mean checking if it contains a section with NOT NOT if I am getting it correctly.

The function should be checking that there is _at least one_ `NOT (NOT (...))` in the circuit structure.  (Is this what you mean by a section with `NOT (NOT (...))` in it?)  This means that it doesn’t matter if there are other instances of `NOT (NOT (...))` elsewhere in the circuit.

~ Matt

**Comment by oflefe** [comment]
Yes, I understand that. However, the line check_NOT ( NOT ( NOT c ) ) = True, only checks if the input to the function check_NOT is exactly, a circuit fed into two consecutive NOT gates; not a circuit that contains two consecutive NOT gates. Is this correct?

**Comment by oflefe** [comment]
Another thing about the proof. When I apply fun.induct and apply auto, sometimes there are a few tasks left. How would I go about continuing the proof and proving the subgoals that are left.

**Comment by MattWindsor91** [comment]
`check_NOT (NOT (NOT c)) = True` looks like a correct case to me, but you'll need to give other cases for `check_NOT` on other circuits.  These will be where you specify how `check_NOT` should work on circuits that aren't a double negation.

For continuing the proof: first, if induction/auto has left any subgoals, I'd double-check to make sure that they make sense (as John mentioned a few lectures ago, sometimes induction proofs can go down the wrong path, and weird subgoals can be a clue).  If so, and if you're using `by` to try to finish the proof using induction/auto, you should be able to replace `by` with `proof`, and then use `case` to specify the remaining cases (@johnwickerson might be able to clarify?).

**Comment by johnwickerson** [comment]
Yes that's right; if you start your proof with something like

```isabelle
theorem "blah"
apply (f.induct)
apply auto
...
```

then you can just carry on this `apply` sequence until you've dispatched all your goals. Or, if you want to use the "structured proof" paradigm then you could start your proof with something like

```isabelle
theorem "blah"
proof (f.induct, auto)
   ...
```

**Comment by Hellonity** [comment]
Hi, I am using induction to prove "check_NOT(opt_NOT c) = False", in the first case I need to prove "check_NOT (opt_NOT c) = False ⟹ check_NOT (opt_NOT (NOT c)) = False". Then Quickcheck found a counterexample for "opt_NOT (NOT c) = NOT (opt_NOT c)"
![1](https://user-images.githubusercontent.com/40735572/70625411-3e68bb80-1c1a-11ea-8fe7-5fac42913260.JPG)
![2](https://user-images.githubusercontent.com/40735572/70625424-40cb1580-1c1a-11ea-9aff-fbf54dea70d1.JPG)

 Isabelle don't know "NOT NOT TRUE" is equal to "TRUE", how can I solve this kind of problem, do I need to simulate the circuit or I am in the wrong direction of proving the theorem?

**Comment by johnwickerson** [comment]
I think this is a similar issue to https://github.com/johnwickerson/HSV/issues/3.

**Comment by Hellonity** [comment]
Thanks, I have checked the issue 3 but I still not understand what does instantiate a theorem mean in this case. Does it mean it tells isabelle to prove the instantiated property?

For example:
![1](https://user-images.githubusercontent.com/40735572/70649319-d11e5000-1c44-11ea-86fc-f769d035a0cf.JPG)

and what should be the type of the instantiation?
![2](https://user-images.githubusercontent.com/40735572/70649497-2d816f80-1c45-11ea-9efb-42f344978679.JPG)

**Comment by MattWindsor91** [comment]
I _think_ the argument to `of` has to be a function from the thing being inducted over to a proposition (from `circuit` to `bool` in this case), and what is being supplied to `of` here is just an implicitly quantified proposition of type `bool`… but I may be mistaken and will need to check next time I’m in front of Isabelle.

**Comment by Hellonity** [comment]
Thanks, the type should be circuit => bool

**Comment by oflefe** [comment]
In my case, it is left with 1 subgoal after auto. And that is 
check_NOT c ⟹ False 
The check_NOT function already defines for all possible c inputs, I do not understand how this even is a subgoal.

**Comment by johnwickerson** [comment]
That subgoal is basically saying "`check_NOT c` is false for any `c`". Which is not true. Which means that you've taken the wrong path earlier in your proof.

**Comment by oflefe** [comment]
Same with Hellonity, I checked issue #3, tried different arguments to "of" but never got it to work.
